### ISSUE TITLE
DynamoDB's maximum item size limit

### ISSUE THREAD

**Issue by adamyy** [castorini/anlessini#11]
DynamoDB imposes a strict item size limit of 400KB, which includes both the attribute name and value. This might be something to worth looking out for.

https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html#limits-items

I actually run into this error when I was trying to import the 2020-09-25 CORD-19 collection.

```
com.amazonaws.services.dynamodbv2.model.AmazonDynamoDBException: Item size has exceeded the maximum allowed size (Service: AmazonDynamoDBv2; Status Code: 400; Error Code: ValidationException; Request ID: TMITLUJHMUQQMO31D0G4FNR4NRVV4KQNSO5AEMVJF66Q9ASUAAJG; Proxy: null)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1811) ~[aws-java-sdk-core-1.11.848.jar:?]
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleServiceErrorResponse(AmazonHttpClient.java:1395) ~[aws-java-sdk-core-1.11.848.jar:?]
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1371) ~[aws-java-sdk-core-1.11.848.jar:?]
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1145) ~[aws-java-sdk-core-1.11.848.jar:?]
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:802) ~[aws-java-sdk-core-1.11.848.jar:?]
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:770) ~[aws-java-sdk-core-1.11.848.jar:?]
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:744) ~[aws-java-sdk-core-1.11.848.jar:?]
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:704) ~[aws-java-sdk-core-1.11.848.jar:?]
        at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:686) ~[aws-java-sdk-core-1.11.848.jar:?]
        at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:550) ~[aws-java-sdk-core-1.11.848.jar:?]
        at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:530) ~[aws-java-sdk-core-1.11.848.jar:?]
        at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.doInvoke(AmazonDynamoDBClient.java:5136) ~[aws-java-sdk-dynamodb-1.11.848.jar:?]
        at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.invoke(AmazonDynamoDBClient.java:5103) ~[aws-java-sdk-dynamodb-1.11.848.jar:?]
        at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.executeBatchWriteItem(AmazonDynamoDBClient.java:691) ~[aws-java-sdk-dynamodb-1.11.848.jar:?]
        at com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.batchWriteItem(AmazonDynamoDBClient.java:657) ~[aws-java-sdk-dynamodb-1.11.848.jar:?]
        at com.amazonaws.services.dynamodbv2.document.internal.BatchWriteItemImpl.doBatchWriteItem(BatchWriteItemImpl.java:113) ~[aws-java-sdk-dynamodb-1.11.848.jar:?]
        at com.amazonaws.services.dynamodbv2.document.internal.BatchWriteItemImpl.batchWriteItem(BatchWriteItemImpl.java:53) ~[aws-java-sdk-dynamodb-1.11.848.jar:?]
        at com.amazonaws.services.dynamodbv2.document.DynamoDB.batchWriteItem(DynamoDB.java:178) ~[aws-java-sdk-dynamodb-1.11.848.jar:?]
        at io.anlessini.utils.ImportCollection$ImporterThread.sendBatchRequest(ImportCollection.java:296) ~[utils-0.1.0-SNAPSHOT.jar:?]
        at io.anlessini.utils.ImportCollection$ImporterThread.run(ImportCollection.java:247) [utils-0.1.0-SNAPSHOT.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
```

Very unfortunately, AWS's [recommendation](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/bp-use-s3-too.html) for circumvent the large item issue is to store the values/items in S3 and store the object identifier in DynamoDB.

Arguably, the largest field of all the collection would be `raw` (more investigation required). Perhaps it would alleviate the issue if we store the document's raw field in S3 (keyed by the docid) and store the rest in DynamoDB.

**Comment by adamyy** [comment]
# Investigation

I wrote a simple [utility](https://github.com/adamyy/anlessini/blob/adamyy%2Fcollection-stats/utils/src/main/java/io/anlessini/utils/CollectionStats.java) to calculate field sizes and tested it on CORD-19 collection (2020-09-25 release).

TL;DR the full text collection has over 300 documents whose size goes beyond 400KB. The biggest one even reached a size of 4MB. The abstract and the paragraphs collection does not have oversized documents.

**Full-text**

```
[ec2-user@ip-172-31-9-72 anlessini]$ utils/target/appassembler/bin/CollectionStats -collection Cord19FullTextCollection -generator Cord19Generator -threads 8 -input ~/collections/cord19-2020-09-25/
(CollectionStats.java:144) - =================== Collection Stats ==================
(CollectionStats.java:147) - Thread pool with 8 threads initialized.
(CollectionStats.java:149) - Initializing collection in /home/ec2-user/collections/cord19-2020-09-25
(CollectionStats.java:152) - 1 file found
(CollectionStats.java:154) - Computing stats...
(CollectionStats.java:165) - 13,863 documents computed
(CollectionStats.java:165) - 21,747 documents computed
(CollectionStats.java:165) - 193,652 documents computed
(CollectionStats.java:165) - 234,206 documents computed
(CollectionStats.java:178) - =================== Collection Stats ==================
(CollectionStats.java:179) - Field                Max(bytes) Min(bytes) Avg(bytes)    Max(KB)    Min(KB)    Avg(KB)
(CollectionStats.java:182) - mag_id                        6          6          6          0          0          0
(CollectionStats.java:182) - author_string            38,992         13         13         38          0          0
(CollectionStats.java:182) - source_x                     35         11         11          0          0          0
(CollectionStats.java:182) - year                          4          4          4          0          0          0
(CollectionStats.java:182) - abstract                 27,019          8          8         26          0          0
(CollectionStats.java:182) - title                     1,199          5          5          1          0          0
(CollectionStats.java:182) - sha                       4,915          3          3          4          0          0
(CollectionStats.java:182) - who_covidence_id             28         16         16          0          0          0
(CollectionStats.java:182) - url                         315          3          3          0          0          0
(CollectionStats.java:182) - pubmed_id                    41          9          9          0          0          0
(CollectionStats.java:182) - license                      18          9          9          0          0          0
(CollectionStats.java:182) - journal                     265          7          7          0          0          0
(CollectionStats.java:182) - contents              4,110,493         11         20      4,014          0          0
(CollectionStats.java:182) - publish_time                 22         12         12          0          0          0
(CollectionStats.java:182) - s2_id                        14          5          5          0          0          0
(CollectionStats.java:182) - id                           10         10         10          0          0          0
(CollectionStats.java:182) - pmcid                        15          5          5          0          0          0
(CollectionStats.java:182) - doi                          71          3          3          0          0          0
(CollectionStats.java:182) - authors                  31,125          7          7         30          0          0
(CollectionStats.java:187) - total                 4,110,799        180        295      4,014          0          0
(CollectionStats.java:190) - Total number of large item: 345
(CollectionStats.java:191) - Large Item docids: [1a3sy8ja, kvztcwu2, 3p81yr4n, dvh0qrgp, 7whai7nr, dwfb81aj, e1bn79ui, w7uhqsio, ka4261wc, ef87c4ej, due9tloa, l36f1gp7, 0s7ok4uw, f1e5cu16, tpqsjjet, 2i96hd5i, eoz1l5fi, lcgeingz, jkke8ije, jqm1hxps, bcw90f5b, zhrjo9w3, rgun9d7k, n8mlxe7p, 89b0m3l5, pgyzluwp, ss1d834y, 75w35fkd, uje7bwrw, ad3i9wgj, d726m8mk, lfeqwzjt, av1gk318, twsy61oq, uemaa8py, mvoq9vln, lnqldz3a, hrxxdqei, 6md4pn3d, rki4yrus, wnk3mmg2, n0kg6jmy, 4057qib7, ezrn8cva, 3bmzc2d3, yg7salqt, w81ysjf9, 5bi0dxuz, v9qgvfll, xhx9pzhj, 2u0ug8po, li8pwigg, ak5pq312, i54q5gsu, 2rzdmh3h, oswgjaxz, ys7an3rr, fqubjnc6, cfhrokfn, x9drxbhk, jrvh9q3c, m9ax1hx5, kck5e1ry, 7yoilsho, xta38e7j, bfbcvd3a, oxzi0b17, 4hxa6vvw, w5xaq43c, 0tgcsa2o, sdqfb591, e1phpqes, c4i9tba2, w790pgz7, get431bj, qxy5kshs, cjd558u7, 2d5lmz8z, pjddd1ft, an9leotw, 6g8p8hzw, kj9520ha, eaqkadyy, pyjkht7u, aqy6q5yl, vv2dh32c, l0otcim5, vgjz74ts, 0fu1h820, olmpzb47, o2e5na78, a3b8hyyr, c3rt4a48, a0jtmuqr, f5w6fw6q, ygeq27dd, h2tmfp3a, jxv4dyz2, 9dgmdtj8, 0xgzynkr, 6d4qtivf, snmf8cvt, jmjq98jy, bie1veti, cgcvfftf, hxj4z228, o36dl4nz, bgbin0y0, 39vjafky, a714injz, i2reanan, 7s38gfq7, 3dg917r0, kqdl8xv3, iu89sx8t, 1mhyo1j6, osstpum6, 14in204v, kb9fnbgy, 69swjzic, ak1umcec, sop7pedo, kjbktnri, oecpqf1j, bnz244jy, rovpo5fj, 5skyttto, uftc8inx, x8ysndmi, t1x9gknd, 47yoz3iy, dzxcfb1u, r0kjp11i, y128d6hb, sizuef1v, ukz4hnmy, q512vvj8, 6bak56zy, zuu359wc, e58sex5y, i2lei4ne, 0y1j99yt, 7r3fjswq, y3o8s7hv, m5n250r2, z0tdpdkk, yoe2bsdo, waus29er, srvqhukl, x90ma8o7, 3pnuj5ru, 598a82iq, nvzfpntu, giw7ic5q, 6qfecv5h, 2y1y8jpx, dhnyu4cw, ri2w5iby, 49jqer1u, o5yjcn7r, roslkxeq, w4t99f62, zi0xpl64, 0am4l5ms, 50ztyq06, xbin9kf5, yppi30a0, dckqft1s, at7pq6u1, syfs8r64, fjh6rqo0, u4n01g04, hdnru3tb, eiogbmxk, wpsxz2h8, pt9q4s8l, 7t5vradw, p6q0f37c, pol2qm74, 2xzc0uc5, sjp9ughi, huc9dyy8, h0o0yqv8, 46fmw0in, zmft62kb, dtp3x41z, y44sfr0c, eitceew2, fwtrahj9, n03zk9ni, ttdtyoqc, qt0worsl, 4gxa9mlm, 2d02eq3y, yknwveyz, 62ap2q4d, i6x5fyvy, iizp0xxm, gf32a6f1, gsyia90r, arfmw07f, ydv1yj4h, z4aaw848, a0qz4tb9, 72cjogxz, m86asgb1, b31ovd5y, yvfjh8xy, 47imh7rj, uj7fe5y6, fxtwkm9j, ox7db2xq, 0t4tiip6, knlwloit, vymlfsdn, zm1nvhwh, 8y5auhou, i10q2569, bhqtip9x, 7gchfffa, yzq9cduh, 4pqabi1d, poamqquy, aei5jyp8, tj3ye1mx, bngoaror, m1cuuehi, mtvuf80b, pnqw7g5p, ojmhkt5o, cmotfjda, n4epxwn9, pgm92fwm, rzzjrs0j, 4zyqqgs0, ja44c3p4, rrverrsj, atolki7k, fgg9s7ou, b9iquizg, 0qy4beiw, ungitgh9, 9wnmdvg5, 9n99l0ig, j9kg00qf, h4k3hkhr, l6wywb5x, kdoht8np, b60knyv7, 9lmudxrh, kdv8t4gq, pcdw8xjf, sfnqz1yi, p8tcv21e, fr55uod6, wobbf76j, phsr10jp, 3zm1j3z5, doj6xgv2, l1jm5vxc, jn20f6xf, rqwobdvt, pyqc6ymz, bdve1ydv, ehq9qnoo, nul6l77m, prwfmov2, 4iaqivp6, dnsdg04n, lv6fy6e6, wkz58d20, sm2o774z, vkcc3bca, bwdfcfh6, r95b9eiz, hpzj1qz7, atbjwpo5, 6le23mn1, 05ybwcat, qc3n0b5k, 2r84ndzv, fio7cjj5, ixyogdmy, t57l3ec7, 4dannjjm, l6c2y4vw, 4508tbhi, ostd0g1w, 0lqlx2rv, tek6p09l, xn11xd78, 79ql1z1d, srxnkgej, un2ysc2v, kt5gt26t, pyv37x5u, 2envy3jm, 3okyyo4k, j58tdxhn, omssacq3, 8sqbqjm1, 0i4q84cf, f2ciho6o, dios78n3, ut56gczm, qm0abgqf, lqi6xegb, iykp3e5s, sqi25kkz, o96m0b9u, 8lwnmdeq, w96h4rir, yjvj6b9b, 578c9ut5, ienpk3uv, aa4iqtuf, 6jpvxlpt, hnf5vayd, y2qap2qm, 3f9yhi09, h9mc51ik, ilhr3iu5, 26su14qs, cvqt35ao, puwg1fjq, 83dahurb, doagajtl, p4o580bz, fed9xg86, m6lkrehi, o2hr5mox, l1obqvsw, xm0cthrv, 0xb64n8p, 6tapkjb6, 3rqnu5bu, 9yiollf6, 2bo0s0hz, rz4r5sj7, 8iwltwig, vf1b8aai, p4vr3067, k1jq4m35, 1s7yfoy2, o1gtp39c]
(CollectionStats.java:193) - ============ Final Counter Values ============
(CollectionStats.java:194) - computed:         279,409
(CollectionStats.java:195) - unindexable:            0
(CollectionStats.java:196) - empty:                 60
(CollectionStats.java:197) - skipped:                6
(CollectionStats.java:198) - errors:                 0
```

**Paragraphs**

```
[ec2-user@ip-172-31-9-72 anlessini]$ utils/target/appassembler/bin/CollectionStats -collection Cord19ParagraphCollection -generator Cord19Generator -threads 8 -input ~/collections/cord19-2020-09-25/
 (CollectionStats.java:144) - =================== Collection Stats ==================
 (CollectionStats.java:147) - Thread pool with 8 threads initialized.
 (CollectionStats.java:149) - Initializing collection in /home/ec2-user/collections/cord19-2020-09-25
 (CollectionStats.java:152) - 1 file found
 (CollectionStats.java:154) - Computing stats...
 (CollectionStats.java:165) - 581,151 documents computed
 (CollectionStats.java:165) - 1,448,793 documents computed
 (CollectionStats.java:165) - 2,145,279 documents computed
 (CollectionStats.java:165) - 2,773,617 documents computed
 (CollectionStats.java:165) - 3,374,916 documents computed
 (CollectionStats.java:178) - =================== Collection Stats ==================
 (CollectionStats.java:179) - Field                Max(bytes) Min(bytes) Avg(bytes)    Max(KB)    Min(KB)    Avg(KB)
 (CollectionStats.java:182) - mag_id                        6          6          6          0          0          0
 (CollectionStats.java:182) - author_string            38,992         13         13         38          0          0
 (CollectionStats.java:182) - source_x                     35         11         11          0          0          0
 (CollectionStats.java:182) - year                          4          4          4          0          0          0
 (CollectionStats.java:182) - abstract                 27,019          8          8         26          0          0
 (CollectionStats.java:182) - title                     1,199          5          5          1          0          0
 (CollectionStats.java:182) - sha                       4,915          3          3          4          0          0
 (CollectionStats.java:182) - who_covidence_id             28         16         16          0          0          0
 (CollectionStats.java:182) - url                         315          3          3          0          0          0
 (CollectionStats.java:182) - pubmed_id                    41          9          9          0          0          0
 (CollectionStats.java:182) - license                      18          9          9          0          0          0
 (CollectionStats.java:182) - journal                     265          7          7          0          0          0
 (CollectionStats.java:182) - contents                188,698         11         11        184          0          0
 (CollectionStats.java:182) - publish_time                 22         12         12          0          0          0
 (CollectionStats.java:182) - s2_id                        14          5          5          0          0          0
 (CollectionStats.java:182) - id                           16         10         10          0          0          0
 (CollectionStats.java:182) - pmcid                        15          5          5          0          0          0
 (CollectionStats.java:182) - doi                          71          3          3          0          0          0
 (CollectionStats.java:182) - authors                  31,125          7          7         30          0          0
 (CollectionStats.java:187) - total                   190,006        180        186        185          0          0
 (CollectionStats.java:190) - Total number of large item: 0
 (CollectionStats.java:191) - Large Item docids: []
 (CollectionStats.java:193) - ============ Final Counter Values ============
 (CollectionStats.java:194) - computed:       3,817,193
 (CollectionStats.java:195) - unindexable:            0
 (CollectionStats.java:196) - empty:                 61
 (CollectionStats.java:197) - skipped:            2,660
 (CollectionStats.java:198) - errors:                 0
```

**Abstract**

```
[ec2-user@ip-172-31-9-72 anlessini]$ utils/target/appassembler/bin/CollectionStats -collection Cord19AbstractCollection -generator Cord19Generator -threads 8 -input ~/collections/cord19-2020-09-25/
(CollectionStats.java:144) - =================== Collection Stats ==================
(CollectionStats.java:147) - Thread pool with 8 threads initialized.
(CollectionStats.java:149) - Initializing collection in /home/ec2-user/collections/cord19-2020-09-25
(CollectionStats.java:152) - 1 file found
(CollectionStats.java:154) - Computing stats...
(CollectionStats.java:165) - 213,237 documents computed
(CollectionStats.java:178) - =================== Collection Stats ==================
(CollectionStats.java:179) - Field                Max(bytes) Min(bytes) Avg(bytes)    Max(KB)    Min(KB)    Avg(KB)
(CollectionStats.java:182) - mag_id                        6          6          6          0          0          0
(CollectionStats.java:182) - author_string            38,992         13         13         38          0          0
(CollectionStats.java:182) - source_x                     35         11         11          0          0          0
(CollectionStats.java:182) - year                          4          4          4          0          0          0
(CollectionStats.java:182) - abstract                 27,019          8          8         26          0          0
(CollectionStats.java:182) - title                     1,199          5          5          1          0          0
(CollectionStats.java:182) - sha                       4,915          3          3          4          0          0
(CollectionStats.java:182) - who_covidence_id             28         16         16          0          0          0
(CollectionStats.java:182) - url                         315          3          3          0          0          0
(CollectionStats.java:182) - pubmed_id                    41          9          9          0          0          0
(CollectionStats.java:182) - license                      18          9          9          0          0          0
(CollectionStats.java:182) - journal                     265          7          7          0          0          0
(CollectionStats.java:182) - contents                 27,172         11         11         26          0          0
(CollectionStats.java:182) - publish_time                 22         12         12          0          0          0
(CollectionStats.java:182) - s2_id                        14          5          5          0          0          0
(CollectionStats.java:182) - id                           10         10         10          0          0          0
(CollectionStats.java:182) - pmcid                        15          5          5          0          0          0
(CollectionStats.java:182) - doi                          71          3          3          0          0          0
(CollectionStats.java:182) - authors                  31,125          7          7         30          0          0
(CollectionStats.java:187) - total                    70,528        180        186         68          0          0
(CollectionStats.java:190) - Total number of large item: 0
(CollectionStats.java:191) - Large Item docids: []
(CollectionStats.java:193) - ============ Final Counter Values ============
(CollectionStats.java:194) - computed:         279,408
(CollectionStats.java:195) - unindexable:            0
(CollectionStats.java:196) - empty:                 61
(CollectionStats.java:197) - skipped:                6
(CollectionStats.java:198) - errors:                 0
```

**Comment by adamyy** [comment]
From the investigation above another thing is also clear: it is not because we stored both `contents` and `raw` that made the overall item size goes beyond limit, it is just that the size of the `contents` were too much for DynamoDB to handle.

**Comment by adamyy** [comment]
Per discussion with @lintool: we should implement a size check before importing the item to DynamoDB, and print warning logs about the outliers.
